### PR TITLE
Users can capture for loops into plxpr using `qml.for_loop`

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
**Context:** The purpose of this PR is to allow `qml.for_loop` to be captured into plxpr.

**Description of the Change:** Now, `qml.for_loop` can be captured into plxpr. The function's behavior resembles the one provided in Catalyst, with the executed branch determined at runtime.

**Benefits:** If `qml.capture.enabled()` is `True`, `qml.for_loop` can be used.

**Possible Drawbacks:** The return type of the loop body function must not depend on the index.

**Related GitHub Issues:** None.

**Related Shortcut Stories** 
[sc-66736]
[sc-70157]

### CODE REVIEW

File: pennylane/compiler/qjit_api.py [hunk 1]
```diff
@@ -568,3 +605,26 @@ def __call__(self, *init_state):
...
 
         return fn_res
+
+    def _call_capture_enabled(self, *init_state):
+
+        import jax  # pylint: disable=import-outside-toplevel
+
+        for_loop_prim = _get_for_loop_qfunc_prim()
+
+        jaxpr_body_fn = jax.make_jaxpr(self.body_fn)
```
Comment (albi3ro) [inline pennylane/compiler/qjit_api.py:615]: So this is a hybrid between the on-the-fly and staged processing?

https://jax.readthedocs.io/en/latest/autodidax.html#on-the-fly-final-style-and-staged-initial-style-processing

It seems like we would neither get the more versatile nature of on-the-fly processing or the straight forward, ahead of time information of staged processing.

I'd be fine with considering use of on-the-fly processing for the control flow primitives, but I'm curious why staged processing doesn't work in this situation.

Comment (PietropaoloFrisoni) [reply pennylane/compiler/qjit_api.py:615]: Considering how `qml.for_loop` works and the corresponding syntax, it seems to me that the most natural approach is to call `jax.make_jaxpr` 'capturing' the body function and then evaluate it with `eval_jaxpr` for every possible value of the index since the index is always the first argument of the body function (I honestly find the syntax of `qml.for_loop` a little counter-intuitive).

However, I am pretty agnostic on the implementation as soon as it does the job, and maybe there are better ways. What would you suggest?

File: pennylane/compiler/qjit_api.py [hunk 2]
```diff
@@ -568,3 +607,28 @@ def __call__(self, *init_state):
...
         return fn_res
+
+    def _call_capture_enabled(self, *init_state):
+
+        import jax  # pylint: disable=import-outside-toplevel
+
+        for_loop_prim = _get_for_loop_qfunc_prim()
+
+        jaxpr_body_fn = jax.make_jaxpr(self.body_fn)(0, *init_state)
+        jaxpr_consts = jax.numpy.array(jaxpr_body_fn.consts)
```
Comment (albi3ro) [inline pennylane/compiler/qjit_api.py:618]: I'm a little confused by this line.  What if the consts are different shapes and types? 

```
x = jax.numpy.array(0.1)
y = jax.numpy.array([1, 2])

@for_loop(0,4,1)
def f(i, a, b):
    return a+x, b+y

f(0.3, jax.numpy.array([5,6]) )
```

I might need to research the effect of casting to an array on jaxpr, but in addition to the shapes/ dtype concern, I also want to be careful about introducing unnecessary copies, as that can lead to excess overhead.

File: pennylane/compiler/qjit_api.py [hunk 3]
```diff
@@ -537,6 +541,41 @@ def _decorator(body_fn):
...
+    def _(lower_bound, upper_bound, step, jaxpr_consts, *init_state, jaxpr_body_fn):
+
+        args = init_state
+        fn_res = args if len(args) > 1 else args[0] if len(args) == 1 else None
+
+        for i in range(lower_bound, upper_bound, step):
+
+            tmp_args = (i, *args)
+            fn_res = jax.core.eval_jaxpr(jaxpr_body_fn.jaxpr, jaxpr_consts, *tmp_args)
+            fn_res = jax.numpy.array(fn_res)
```
Comment (albi3ro) [inline pennylane/compiler/qjit_api.py:563]: Same concern as raised below.  What if the outputs have different types and shapes? Why is this needed?

File: pennylane/compiler/qjit_api.py [hunk 4]
```diff
@@ -537,6 +541,41 @@ def _decorator(body_fn):
...
+            args = fn_res if len(args) > 1 else (fn_res,) if len(args) == 1 else ()
```
Comment (albi3ro) [inline pennylane/compiler/qjit_api.py:564]: So we know that `jax.core.eval_jaxpr` *always* returns a list of tensorlikes. So I'm little unsure what this line is doing.

What cases give rise to each logical branch?

Since `fn_res` is already a list, why do we want `([a, ], )`?

Comment (albi3ro) [reply pennylane/compiler/qjit_api.py:564]: It might be worth breaking this out into multi-line form:
```
if len(args) > 1:
    args = fn_res
elif len(args) == 1:
    args = (fn_res,)
else:
   args = ()
```

And making sure we actually have test coverage for each line.

File: pennylane/compiler/qjit_api.py [hunk 5]
```diff
@@ -537,6 +541,41 @@ def _decorator(body_fn):
...
+    import jax  # pylint: disable=import-outside-toplevel
+
+    for_loop_prim = jax.core.Primitive("for_loop")
+    for_loop_prim.multiple_results = True
+
+    @for_loop_prim.def_impl
+    def _(lower_bound, upper_bound, step, jaxpr_consts, *init_state, jaxpr_body_fn):
+
+        args = init_state
+        fn_res = args if len(args) > 1 else args[0] if len(args) == 1 else None
```
Comment (albi3ro) [inline pennylane/compiler/qjit_api.py:557]: Do we need this line?

File: pennylane/compiler/qjit_api.py (hunk 2 above)
Comment (PietropaoloFrisoni) [reply pennylane/compiler/qjit_api.py:618]: Thanks, please note that this is not the final implementation and there are several improvements/changes/tests required (the PR is still a draft)

File: doc/releases/changelog-dev.md
Comment (PietropaoloFrisoni) [inline doc/releases/changelog-dev.md:1]: @albi3ro Should we combine all changelog entries for program capture into a unique entry?

File: pennylane/compiler/qjit_api.py [hunk 6]
```diff
@@ -537,6 +541,37 @@ def _decorator(body_fn):
...
+    def _(lower_bound, upper_bound, step, *jaxpr_consts_and_init_state, jaxpr_body_fn, n_consts):
+
+        jaxpr_consts = jaxpr_consts_and_init_state[:n_consts]
+        init_state = jaxpr_consts_and_init_state[n_consts:]
+
+        # in case lower_bound >= upper_bound, return the initial state
+        fn_res = init_state
+
+        for i in range(lower_bound, upper_bound, step):
+            fn_res = jax.core.eval_jaxpr(jaxpr_body_fn.jaxpr, jaxpr_consts, *(i, *fn_res))
```
Comment (albi3ro) [inline pennylane/compiler/qjit_api.py:563]:
```suggestion
            fn_res = jax.core.eval_jaxpr(jaxpr_body_fn.jaxpr, jaxpr_consts, i, *fn_res)
```

Comment (PietropaoloFrisoni) [reply pennylane/compiler/qjit_api.py]: Cool, I thought this didn't work

File: pennylane/compiler/qjit_api.py [hunk 7]
```diff
@@ -568,3 +603,28 @@ def __call__(self, *init_state):
...
 
         return fn_res
+
+    def _call_capture_enabled(self, *init_state):
+
+        import jax  # pylint: disable=import-outside-toplevel
+
+        for_loop_prim = _get_for_loop_qfunc_prim()
+
+        jaxpr_body_fn = jax.make_jaxpr(self.body_fn)(0, *init_state)
```
Comment (obliviateandsurrender) [inline pennylane/compiler/qjit_api.py:613]: What is the first argument `0` for `jax.make_jaxpr(self.body_fn)` here for? 😅

File: pennylane/compiler/qjit_api.py [hunk 8]
```diff
@@ -537,6 +541,37 @@ def _decorator(body_fn):
     return _decorator
 
 
+@functools.lru_cache
+def _get_for_loop_qfunc_prim():
+    """Get the loop_for primitive for quantum functions."""
+
+    import jax  # pylint: disable=import-outside-toplevel
```
Comment (obliviateandsurrender) [inline pennylane/compiler/qjit_api.py:548]: Importing `jax` here wouldn't add an overhead? 🤔

File: pennylane/compiler/qjit_api.py (hunk 7 above)
Comment (PietropaoloFrisoni) [reply pennylane/compiler/qjit_api.py:613]: This is because of how the syntax of `qml.for` works. This `0` corresponds to the index of the for loop (it does not matter the exact value as long as it is an integer since the value is established at runtime). By convention, it always corresponds to the first argument of the loop body function, and we must capture it here along with the other arguments.

File: pennylane/compiler/qjit_api.py (hunk 8 above)
Comment (PietropaoloFrisoni) [reply pennylane/compiler/qjit_api.py:548]: We must use `jax` functions within `_get_for_loop_qfunc_prim`. We used this scheme for all the previous implementations, but (although I am not 100% sure) I think the overhead does almost not exist in this case since when we reach this part of the code `jax` should have been imported already

Comment (albi3ro) [reply pennylane/compiler/qjit_api.py:548]: We can't import jax outside this function, since that would cause `import pennylane as qml` to fail if jax is not installed.  We have to hide the dependency and make the jax import a bit lazy. This primitive is only defined "as needed", instead of "on import".

But we don't want to define the primitive a bunch of times with additional overhead, so we slapped the `lru_cache` around this constructor.

That makes it so the primitive is created once the first time it's needed, and then just retrieved every other time.

File: pennylane/compiler/qjit_api.py (hunk 7 above)
Comment (albi3ro) [reply pennylane/compiler/qjit_api.py:613]: `0` forces the creation of the jaxpr with the first argument being a single integer. We could equivalently use any other integer, or `self.lower_bound`, but using `0` is a simple enough way to construct the correct jaxpr.

File: tests/capture/test_capture_for_loop.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+                return a
+
+            a2 = loop(a)
+            return a2
+
+        expected = jax.numpy.ones(arg)
+        result = fn(arg)
+        assert np.allclose(result, expected), f"Expected {expected}, but got {result}"
+
+        # Note that this cannot be transformed with `jax.make_jaxpr`
```
Comment (obliviateandsurrender) [inline tests/capture/test_capture_for_loop.py:63]: Maybe you could do something like this to use `jax.make_jaxpr`

```py
def fn(arg):

    a = jax.numpy.ones(arg.shape)

    @qml.for_loop(0, 10, 2)
    def loop(_, a):
        return a

    a2 = loop(a)
    return a2
```
```py
>>> jax.make_jaxpr(fn)(jnp.zeros(2)), fn(jnp.zeros(2))
({ lambda ; a:f32[2]. let
    b:f32[2] = broadcast_in_dim[broadcast_dimensions=() shape=(2,)] 1.0
    c:f32[2] = for_loop[
      jaxpr_body_fn={ lambda ; d:i32[] e:f32[2]. let  in (e,) }
      n_consts=0
    ] 0 10 2 b
  in (c,) },
 [Array([1., 1.], dtype=float32)])
```

File: tests/capture/test_capture_for_loop.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+            @qml.for_loop(0, 10, 1)
+            def loop_body(i, x):
+                qml.RX(arg1, wires=0)
+                qml.RX(arg2, wires=0)
+                qml.RY(jax.numpy.sin(x), wires=0)
+                return x + i**2
+
+            loop_body(arg)
+
+            return qml.expval(qml.PauliZ(0))
```
Comment (obliviateandsurrender) [inline tests/capture/test_capture_for_loop.py:214]:
```suggestion
            return qml.expval(qml.Z(0))
```

File: tests/capture/test_capture_for_loop.py (hunk 9 above)
Comment (PietropaoloFrisoni) [reply tests/capture/test_capture_for_loop.py:63]: Yes, but I don't think it's crucial. If I remember correctly, I took this test from Catalyst and I only wanted to ensure that the implementation of the primitive for `qml.for` worked as expected for such a case

File: pennylane/compiler/qjit_api.py (hunk 7 above)
Comment (obliviateandsurrender) [reply pennylane/compiler/qjit_api.py:613]: Ahh okay, I was just confused about the choice for using `0` and not `self.lower_bound`. Thanks for clarifying!

**Review by obliviateandsurrender — APPROVED** [review]
Thanks @PietropaoloFrisoni, looks good to me! 🚀

**Review by albi3ro — APPROVED** [review]
A bit hesitant about the hardcoded expected results in the tests, but I don't think that concern is blocking the approval of the PR.

🎉  🚀 excellent job:)
